_posts/2022-09-XX-rate-limiting-web-applications.markdown

```diff
@@ -0,0 +1,465 @@
+---
+layout: post
+title: "Rate limiting in web applications - Concepts and approaches"
+date: 2022-08-XX 03:44:05 +0100
+comments: true
+published: true
+categories: ["post"]
+tags: ["General", "Web"]
+author: Maarten Balliauw
+---
+
+Your web application is running fine, and your users are behaving as expected. Life is good!
+
+Is it, though...? Users are probably using your application in ways you did not expect.
+Crazy usage patterns resulting in more requests than expected, request bursts when users come back to the office after the weekend, and more!
+
+{% include toc %}
+
+TODO this post, plus a practical sample
+
+These unexpected requests all pose a potential threat to the health of your web application and may impact other users or the service as a whole.
+Ideally, you want to put a bouncer at the door to do some filtering: limit the number of requests over a given timespan, limiting bandwidth, ...
+
+In this post, let's explore the simple yet wide realm of rate limiting.
+We'll go over how to decide which resources to limit, what these limits should be, and where to enforce these limits.
+
+As a (mostly) .NET developer myself, I'll use some examples and link some resources that use ASP.NET Core.
+The general concepts however will also apply to other platforms and web frameworks.
+
+## Introduction to rate limiting
+
+Before we dive into the details, let's start with an introduction about why you would want to apply rate limiting, and what it is.
+
+### Why rate limiting?
+
+Let's say you are building a web API that lets you store todo items.
+Nice and simple: a `GET /api/todos` that returns a list of todo items, and a `POST /api/todos` that let you create and update a specific todo item.
+What could possibly go wrong with using these two endpoints?
+
+Off the top of my head:
```

Comment:
You depend on an external system that throttles you for X requests, so let's stop that from happening and stop the user from triggering that